applications/OptimizationApplication/custom_utilities/filtering/explicit_vertex_morphing_filter.cpp

```diff
@@ -0,0 +1,311 @@
+    const auto& r_origin_expression = rContainerExpression.GetExpression();
+    const auto& r_container = rContainerExpression.GetContainer();
+    const auto& r_filter_radius_expression = mpFilterRadiusContainer->GetExpression();
+
+    ContainerExpression<TContainerType> result(*rContainerExpression.pGetModelPart());
+    auto p_expression = LiteralFlatExpression<double>::Create(result.GetContainer().size(), rContainerExpression.GetItemShape());
+    result.SetExpression(p_expression);
+
+    struct TLS
+    {
+        explicit TLS(const IndexType MaxNumberOfNeighbors)
+        {
+            mNeighbourEntityPoints.resize(MaxNumberOfNeighbors);
+            mResultingSquaredDistances.resize(MaxNumberOfNeighbors);
+        }
+
+        EntityPointVector mNeighbourEntityPoints;
+        std::vector<double> mResultingSquaredDistances;
+    };
+
+    IndexPartition<IndexType>(r_container.size()).for_each(TLS(mMaxNumberOfNeighbors), [&](const IndexType Index, TLS& rTLS){
+        const double vertex_morphing_radius = r_filter_radius_expression.Evaluate(Index, Index, 0);
+
+        EntityPoint<EntityType> entity_point(*(r_container.begin() + Index), Index);
+        const auto number_of_neighbors = mpSearchTree->SearchInRadius(
+                                            entity_point,
+                                            vertex_morphing_radius,
+                                            rTLS.mNeighbourEntityPoints.begin(),
+                                            rTLS.mResultingSquaredDistances.begin(),
+                                            mMaxNumberOfNeighbors);
+
+        std::vector<double> list_of_weights(number_of_neighbors, 0.0);
+        double sum_of_weights = 0.0;
+        ExplicitVertexMorphingFilterHelperUtilities::ComputeWeightForAllNeighbors(
+            sum_of_weights, list_of_weights, *mpKernelFunction, vertex_morphing_radius,
+            entity_point, rTLS.mNeighbourEntityPoints, number_of_neighbors);
+
+        const IndexType current_data_begin = Index * stride;
+
+        for (IndexType j = 0; j < stride; ++j) {
+            double& current_index_value = *(p_expression->begin() + current_data_begin + j);
+            current_index_value = 0.0;
+
+            for(IndexType neighbour_index = 0 ; neighbour_index < number_of_neighbors; ++neighbour_index) {
+                const IndexType neighbour_id = rTLS.mNeighbourEntityPoints[neighbour_index]->Id();
+                const double weight = list_of_weights[neighbour_index] / sum_of_weights;
+                const double origin_value = r_origin_expression.Evaluate(neighbour_id, neighbour_id * stride, j);
+                current_index_value +=  weight * origin_value;
+            }
+        }
+    });
+
+    return result;
+
+    KRATOS_CATCH("");
+}
+
+template<class TContainerType>
+ContainerExpression<TContainerType> ExplicitVertexMorphingFilter<TContainerType>::UnFilterField(const ContainerExpression<TContainerType>& rContainerExpression) const
```

Comment:
I think here MapBackward is misunderstood as unfiltering a field. Please note that Unfilter field for explicit filtering requires solution of a linear system of equations.

Reply:
What does UnFilterField mean? A^-1 ? In VM we don't unfilter anything as far as I know ...

Reply:
Removed.